### PR TITLE
Switch to use Web API in Slidedown and Tagging Container

### PR DESCRIPTION
**Pull request by rgomezp**
# Description
This change includes a transition from using hard-coded HTML strings to using the W3C Web API to create DOM elements programmatically. The main reason for this is for security purposes since this approach is an easy way to avoid injection vulnerabilities.

This PR also includes other tweaks like the addition of CSS and Element ID constants, 1 file rename, and other minor nits.

# Systems Affected
WebSDK only

# Validation

### Tests
   - [x] All the automated tests pass or I explained why that is not possible
   - [x] I have personally tested this on my machine or explained why that is not possible
   - [x] I have included test coverage for these changes or explained why they are not needed

Test coverage for these changes are not needed since the only change was a switch to how DOM elements are created.

### Screenshots
   - [x] I have included screenshots/recordings of the intended results or explained why they are not needed

Nothing has visually changed in this PR.

---

This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/681)

### CODE REVIEW

File: src/slidedown/Slidedown.ts [hunk 1]
```diff
@@ -127,83 +135,92 @@ export default class Slidedown {
...
       Log.debug("Slidedown private category options are not defined");
       return;
     }
 
     if (state) {
       // note: savingButtonText is hardcoded in constructor. TODO: pull from config & set defaults for future release
-      this.allowButton.innerHTML = this.getIndicatorHolderHtmlWithText(this.categoryOptions!.savingButtonText!);
-      addDomElement(this.buttonIndicatorHolder, 'beforeend',
-        getLoadingIndicatorWithColor(COLORS.whiteLoadingIndicator));
+      this.allowButton.innerHTML = this.getIndicatorHolderHtmlWithText(this.categoryOptions.savingButtonText).innerHTML;
```
Comment (jkasten2) [inline src/slidedown/Slidedown.ts:146]: nit, use [`replaceChild`](https://developer.mozilla.org/en-US/docs/Web/API/Node/replaceChild) over `innerHTML`.

File: src/slidedown/Slidedown.ts [hunk 2]
```diff
@@ -127,83 +135,92 @@ export default class Slidedown {
...
+  setFailureState(state: boolean): void {
     if (!this.categoryOptions) {
       Log.debug("Slidedown private category options are not defined");
       return;
     }
 
     if (state) {
       // note: errorButtonText is hardcoded in constructor. TODO: pull from config & set defaults for future release
-      this.allowButton.innerHTML = this.getIndicatorHolderHtmlWithText(this.categoryOptions!.errorButtonText);
+      this.allowButton.innerHTML = this.getIndicatorHolderHtmlWithText(this.categoryOptions.errorButtonText).innerHTML;
```
Comment (jkasten2) [inline src/slidedown/Slidedown.ts:169]: nit, use [`replaceChild`](https://developer.mozilla.org/en-US/docs/Web/API/Node/replaceChild) over `innerHTML`.

File: src/slidedown/TaggingContainer.ts [hunk 3]
```diff
@@ -8,102 +8,135 @@ import {
...
+        const wrappingDiv   = document.createElement("div");
+
+        addCssClass(labelElement, TAGGING_CONTAINER_CSS_CLASSES.categoryLabel);
+        addCssClass(labelSpan, TAGGING_CONTAINER_CSS_CLASSES.categoryLabelText);
+        addCssClass(inputElement, TAGGING_CONTAINER_CSS_CLASSES.categoryLabelInput);
+        addCssClass(checkmarkSpan, TAGGING_CONTAINER_CSS_CLASSES.checkmark);
+
+        labelElement.title = label;
+        labelSpan.innerText = label;
+        inputElement.type    = "checkbox";
```
Comment (jkasten2) [inline src/slidedown/TaggingContainer.ts:102]: nit,  inconsistent spaces on these lines

File: test/support/sdk/TestEnvironment.ts [hunk 4]
```diff
@@ -221,13 +220,14 @@ export class TestEnvironment {
     }
 
     let html = '<!doctype html><html><head></head><body></body></html>';
+    global.document = jsdom.jsdom('');
```
Comment (jkasten2) [inline test/support/sdk/TestEnvironment.ts:223]: Why was this line required?

Comment (rgomezp) [reply test/support/sdk/TestEnvironment.ts:223]: the global document object must be defined for `getSlidedownElement` to work correctly (below). this line initializes the document object so that function works.

Comment (jkasten2) [reply test/support/sdk/TestEnvironment.ts:223]: ah, gotcha 👍

File: src/slidedown/Slidedown.ts (hunk 2 above)
Comment (jkasten2) [reply src/slidedown/Slidedown.ts:169]: Looking at this again [`replaceWith`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/replaceWith) would probably be a better fit in this case.

File: src/slidedown/Slidedown.ts (hunk 1 above)
Comment (jkasten2) [reply src/slidedown/Slidedown.ts:146]: Looking at this again [`replaceWith`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/replaceWith) would probably be a better fit in this case.

Comment (rgomezp) [reply src/slidedown/Slidedown.ts:146]: Ah, yep that makes more sense

File: src/slidedown/Slidedown.ts (hunk 2 above)
Comment (rgomezp) [reply src/slidedown/Slidedown.ts:169]: Ah, yep that makes more sense

File: src/slidedown/Slidedown.ts (hunk 1 above)
Comment (rgomezp) [reply src/slidedown/Slidedown.ts]: I ended up switching to use `insertAdjacentElement` ([mdn docs](https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentElement)) to individually insert the children nodes. I had to also separate the helper function into two to return the span element and indicator holder separately.

File: src/slidedown/Slidedown.ts (hunk 2 above)
Comment (rgomezp) [reply src/slidedown/Slidedown.ts]: I ended up switching to use `insertAdjacentElement` ([mdn docs](https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentElement)) to individually insert the children nodes. I had to also separate the helper function into two to return the span element and indicator holder separately.

File: src/slidedown/Slidedown.ts [hunk 5]
```diff
@@ -143,14 +143,18 @@ export default class Slidedown {
 
     if (state) {
       // note: savingButtonText is hardcoded in constructor. TODO: pull from config & set defaults for future release
-      this.allowButton.innerHTML = this.getIndicatorHolderHtmlWithText(this.categoryOptions.savingButtonText).innerHTML;
       this.allowButton.disabled = true;
+      this.allowButton.textContent = null;
```
Comment (jkasten2) [inline src/slidedown/Slidedown.ts:147]: Is it necessary to remove the `textContent`?

Comment (rgomezp) [reply src/slidedown/Slidedown.ts:147]: Yeah, otherwise it will just get appended to the end and we'll have the previous text plus the new text and the spinning indicator

Comment (jkasten2) [reply src/slidedown/Slidedown.ts:147]: 👍 ah gotcha

**Comment by rgomezp** [comment]
- [x] Visually normal slidedown.
- [x] Visually category slidedown.
- [x] Saving categories saves correct values
- [x] Visually category slidedown when updating.
